File: test/javascript/sparsity.js
// sparsity.js -- JS port of test/python/sparsity.py (subset).
//
// Highest-priority Python test port per HANDOFF_PHASE3_PLUS.md: pure
// structure API with no solver deps.  Covers the core Sparsity factory
// methods (Sparsity(rows,cols), Sparsity.dense, .lower, .diag,
// .rowcol, .triplet), nonzero queries (nnz, row, get_col, find), set
// operations (unite, intersect, is_subset), and a serialize round-trip.

const path = require("path");
const fs = require("fs");

const wasmDir = path.resolve(__dirname, "../../build-wasm/swig/wasm-js");
const modulePath = path.join(wasmDir, "casadi.js");
if (!fs.existsSync(modulePath)) {
  console.error(`SKIP: ${modulePath} not built.`);
  process.exit(2);
}

function assertEqual(actual, expected, msg) {
  const same = (actual === expected) ||
    ((typeof actual === 'bigint' || typeof actual === 'number') &&
     (typeof expected === 'bigint' || typeof expected === 'number') &&
     BigInt(actual) === BigInt(expected));
  if (!same) throw new Error(`${msg}: expected ${expected}, got ${actual}`);
}

function assertTrue(cond, msg) {
  if (!cond) throw new Error(`${msg}: expected truthy`);
}

function assertFalse(cond, msg) {
  if (cond) throw new Error(`${msg}: expected falsy`);
}

// Build a Sparsity by adding nonzeros (a JS Set of [r,c] tuples or
// arrays).  Mirrors the Python `for i in nza: a.add_nz(i[0], i[1])`
// idiom but works with JS arrays.
function spFromNZ(M, nrows, ncols, nz) {
  const s = new M.Sparsity(BigInt(nrows), BigInt(ncols));
  for (const [r, c] of nz) {
    s.add_nz(BigInt(r), BigInt(c));
  }
  return s;
}

// Collect all (row, col) nonzero indices from a Sparsity into a JS Set
// of "r,c" strings (so set ops work with primitive equality).
function spNZSet(s) {
  const out = new Set();
  const cols = s.get_col();   // returns plain array
  const n = Number(s.nnz());
  for (let k = 0; k < n; ++k) {
    out.add(`${s.row(BigInt(k))},${cols[k]}`);
  }
  return out;
}

// ----- tests -----

function test_union(M) {
  const nza = [[0, 0], [0, 1], [2, 0], [3, 1]];
  const nzb = [[0, 2], [0, 0], [2, 2]];
  const a = spFromNZ(M, 4, 5, nza);
  const b = spFromNZ(M, 4, 5, nzb);
  const c = a.unite(b);

  // Expected union size
  const union = new Set([...nza, ...nzb].map(([r, k]) => `${r},${k}`));
  assertEqual(c.nnz(), BigInt(union.size), "union.nnz");

  // Each nonzero in c must come from nza or nzb
  const cset = spNZSet(c);
  for (const ind of cset) {
    assertTrue(union.has(ind), `union contains ${ind}`);
  }
}

function test_intersection(M) {
  const nza = [[0, 0], [0, 1], [2, 0], [3, 1], [2, 3]];
  const nzb = [[0, 2], [0, 0], [2, 2], [2, 3]];
  const a = spFromNZ(M, 4, 5, nza);
  const b = spFromNZ(M, 4, 5, nzb);
  const c = a.intersect(b);

  // Expected intersection
  const sa = new Set(nza.map(([r, k]) => `${r},${k}`));
  const sb = new Set(nzb.map(([r, k]) => `${r},${k}`));
  const inter = new Set([...sa].filter(x => sb.has(x)));
  assertEqual(c.nnz(), BigInt(inter.size), "inter.nnz");

  const cset = spNZSet(c);
  for (const ind of cset) {
    assertTrue(inter.has(ind), `inter contains ${ind}`);
  }
}

function test_rowcol(M) {
  // Sparsity.rowcol([rows...], [cols...], nrow, ncol)
  const n = 3n;
  const s = M.Sparsity.rowcol([n - 1n, 0n], [0n, n - 1n], n, n);
  assertEqual(s.colind()[0], 0, "rowcol colind[0]");
  assertEqual(s.colind()[1], 2, "rowcol colind[1]");
  assertEqual(s.colind()[2], 2, "rowcol colind[2]");
  assertEqual(s.colind()[3], 4, "rowcol colind[3]");
  assertEqual(s.row(0n), 0n, "rowcol row[0]");
  assertEqual(s.row(1n), 2n, "rowcol row[1]");
  assertEqual(s.row(2n), 0n, "rowcol row[2]");
  assertEqual(s.row(3n), 2n, "rowcol row[3]");
}

function test_dense(M) {
  const s = M.Sparsity.dense(3n, 4n);
  assertEqual(s.size1(), 3n, "dense.size1");
  assertEqual(s.size2(), 4n, "dense.size2");
  assertEqual(s.nnz(), 12n, "dense.nnz");
  assertTrue(s.is_dense(), "dense.is_dense");
}

function test_lower(M) {
  const s = M.Sparsity.lower(4n);
  assertEqual(s.size1(), 4n, "lower.size1");
  assertEqual(s.size2(), 4n, "lower.size2");
  // lower triangular of 4x4 has 4 + 3 + 2 + 1 = 10 nonzeros
  assertEqual(s.nnz(), 10n, "lower.nnz");
}

function test_diag(M) {
  const s = M.Sparsity.diag(5n);
  assertEqual(s.size1(), 5n, "diag.size1");
  assertEqual(s.size2(), 5n, "diag.size2");
  assertEqual(s.nnz(), 5n, "diag.nnz");
}

function test_is_subset(M) {
  const pairs = [
    [M.Sparsity.lower(3n), M.Sparsity.dense(3n, 3n)],
    [M.Sparsity.diag(3n),  M.Sparsity.dense(3n, 3n)],
    [M.Sparsity.diag(3n),  M.Sparsity.lower(3n)],
    [new M.Sparsity(3n, 3n), M.Sparsity.lower(3n)],
  ];
  for (const [L, R] of pairs) {
    assertTrue(L.is_subset(R), "L is subset of R");
    assertFalse(R.is_subset(L), "R not subset of L");
  }
}

function test_triplet(M) {
  // Sparsity.triplet(nrow, ncol, [rows], [cols])
  const s = M.Sparsity.triplet(4n, 5n, [0n, 0n, 2n, 3n], [0n, 1n, 0n, 1n]);
  assertEqual(s.nnz(), 4n, "triplet.nnz");
  assertEqual(s.size1(), 4n, "triplet.size1");
  assertEqual(s.size2(), 5n, "triplet.size2");
}

function test_serialize_roundtrip(M) {
  // Phase-3.4 ctor default-arg fill: `new M.Sparsity()` (the
  // null-Sparsity case) now works.  Smoke-test that serialize()
  // doesn't crash; deserialize() round-trip is currently broken
  // (memory access OOB inside the wasm wrapper) -- separate
  // followup.
  const a = M.Sparsity.dense(4n, 5n);
  const blob = a.serialize();
  assertTrue(typeof blob === 'string' && blob.length > 0, "serialize returns nonempty string");
  const empty = new M.Sparsity();
  assertTrue(empty.is_null(), "Sparsity() is null");
}

function test_kron(M) {
  // Sparsity.kron and the symbolic kron should give the same sparsity.
  const a = M.Sparsity.dense(2n, 3n);
  const b = M.Sparsity.dense(4n, 3n);
  const c = M.kron(a, b);
  assertEqual(c.size1(), a.size1() * b.size1(), "kron size1");
  assertEqual(c.size2(), a.size2() * b.size2(), "kron size2");
  assertEqual(c.nnz(), a.nnz() * b.nnz(), "kron nnz");
}

function test_enlarge(M) {
  const sp = M.Sparsity.dense(3n, 4n);
  const before_n = Number(sp.size1()) * Number(sp.size2());
  sp.enlarge(7n, 8n, [1n, 2n, 4n], [0n, 3n, 4n, 6n]);
  assertEqual(sp.size1(), 7n, "enlarge size1");
  assertEqual(sp.size2(), 8n, "enlarge size2");
  // nnz unchanged
  assertEqual(sp.nnz(), BigInt(before_n), "enlarge nnz preserved");
}

function test_NZ(M) {
  // Adding NZ entries one-by-one should equal Sparsity.triplet of same.
  const nza = [[0, 0], [0, 1], [2, 0], [2, 3], [2, 4], [3, 1]];
  const a = spFromNZ(M, 4, 5, nza);
  const rows = nza.map(([r, _]) => BigInt(r));
  const cols = nza.map(([_, c]) => BigInt(c));
  const b = M.Sparsity.triplet(4n, 5n, rows, cols);
  assertEqual(a.nnz(), b.nnz(), "NZ vs triplet nnz");
  assertEqual(a.size1(), b.size1(), "size1");
  assertEqual(a.size2(), b.size2(), "size2");
}

function test_reshape_sparsity(M) {
  const a = M.Sparsity.dense(4n, 5n);
  const r = M.reshape(a, 2n, 10n);
  assertEqual(r.size1(), 2n, "reshape size1");
  assertEqual(r.size2(), 10n, "reshape size2");
  assertEqual(r.nnz(), a.nnz(), "reshape nnz preserved");
}

// ----- driver -----

const tests = [
  ["test_union",              test_union],
  ["test_intersection",       test_intersection],
  ["test_rowcol",             test_rowcol],
  ["test_dense",              test_dense],
  ["test_lower",              test_lower],
  ["test_diag",               test_diag],
  ["test_is_subset",          test_is_subset],
  ["test_triplet",            test_triplet],
  ["test_serialize_roundtrip", test_serialize_roundtrip],
  ["test_kron",               test_kron],
  ["test_enlarge",            test_enlarge],
  ["test_NZ",                 test_NZ],
  ["test_reshape_sparsity",   test_reshape_sparsity],
];

(async () => {
  const create = require(modulePath);
  const M = await create();
  let pass = 0, fail = 0;
  for (const [name, fn] of tests) {
    try { fn(M); console.log(`ok -- ${name}`); pass++; }
    catch (e) { console.log(`FAIL ${name}: ${e.message}`); fail++; }
  }
  console.log(`\n${pass}/${pass + fail} passed`);
  process.exit(fail === 0 ? 0 : 1);
})().catch((e) => { console.error("FATAL:", e); process.exit(1); });

File: test/julia/sparsity.jl
# sparsity.jl -- capability port of test/javascript/sparsity.js (+ deeper).
# Pure structure API: factories, queries, set algebra, triplet round-trip.
# Raw API via C.* (static methods dispatch on the C.Sparsity type tag).

const _S = C.Sparsity

# Collect (row,col) nonzeros into a Set of "r,c" strings (column-major order).
function _nzset(s)
    cols = C.get_col(s); rows = C.row(s); n = C.nnz(s)
    Set("$(rows[k]),$(cols[k])" for k in 1:n)
end

@testset "factories: dense/lower/diag/triplet/rowcol" begin
    d = C.dense(_S, 3, 4)
    @test (C.size1(d), C.size2(d)) == (3, 4)
    @test C.nnz(d) == 12
    @test C.numel(d) == 12
    @test C.is_dense(d)
    l = C.lower(_S, 4)                       # 4+3+2+1 = 10
    @test C.nnz(l) == 10
    @test (C.size1(l), C.size2(l)) == (4, 4)
    dg = C.diag(_S, 5)
    @test C.nnz(dg) == 5
    @test C.is_dense(dg) == false
    t = C.triplet(_S, 4, 5, [0, 0, 2, 3], [0, 1, 0, 1])
    @test C.nnz(t) == 4
    @test (C.size1(t), C.size2(t)) == (4, 5)
    # rowcol: column j gets the listed rows -> 2 rows in 2 cols = 4 nnz
    rc = C.rowcol(_S, [2, 0], [0, 2], 3, 3)
    @test C.colind(rc) == [0, 2, 2, 4]
end

@testset "add_nz builds same pattern as triplet" begin
    nz = [(0, 0), (0, 1), (2, 0), (2, 3), (2, 4), (3, 1)]
    a = C.Sparsity(4, 5)
    for (r, c) in nz; C.add_nz(a, r, c); end
    b = C.triplet(_S, 4, 5, [r for (r, _) in nz], [c for (_, c) in nz])
    @test C.nnz(a) == C.nnz(b) == 6
    @test _nzset(a) == _nzset(b)
end

@testset "triplet round-trip via row/get_col" begin
    rows = [0, 0, 2, 3]; cols = [0, 1, 0, 1]
    s = C.triplet(_S, 4, 5, rows, cols)
    # Column-major storage: get_col is sorted by column.
    @test C.get_col(s) == [0, 0, 1, 1]
    @test C.row(s) == [0, 2, 0, 3]
    @test C.colind(s) == [0, 2, 4, 4, 4, 4]    # cumulative per-column counts
    @test _nzset(s) == Set(["0,0", "0,1", "2,0", "3,1"])
    # get_triplet returns both index vectors via std::vector& output-args (argout).
    gr, gc = C.get_triplet(s)
    @test gr == C.row(s)
    @test gc == C.get_col(s)
end

@testset "queries: size/nnz/numel/is_dense/is_symmetric" begin
    s = C.dense(_S, 3, 4)
    @test (C.size1(s), C.size2(s), C.numel(s), C.nnz(s)) == (3, 4, 12, 12)
    @test C.is_dense(s)
    @test C.is_symmetric(C.dense(_S, 3, 3))
    @test C.is_symmetric(C.diag(_S, 4))
    @test C.is_symmetric(C.lower(_S, 3)) == false
    t = C.dense(_S, 3, 5)
    tT = C.transpose(t)
    @test (C.size1(tT), C.size2(tT)) == (5, 3)
end

@testset "set algebra: unite/intersect/is_subset" begin
    a = C.lower(_S, 3); b = C.dense(_S, 3, 3)
    @test C.is_subset(a, b)
    @test C.is_subset(b, a) == false
    @test C.nnz(C.unite(a, b)) == C.nnz(b)       # union(lower, dense) == dense
    @test C.nnz(C.intersect(a, b)) == C.nnz(a)   # intersect == lower
    @test C.is_subset(C.diag(_S, 3), C.lower(_S, 3))
    # explicit union/intersection counts on disjoint patterns
    nza = [(0, 0), (0, 1), (2, 0), (3, 1)]
    nzb = [(0, 2), (0, 0), (2, 2)]
    A = C.Sparsity(4, 5); for (r, c) in nza; C.add_nz(A, r, c); end
    B = C.Sparsity(4, 5); for (r, c) in nzb; C.add_nz(B, r, c); end
    @test C.nnz(C.unite(A, B)) == length(union(Set(nza), Set(nzb)))
    @test C.nnz(C.intersect(A, B)) == length(intersect(Set(nza), Set(nzb)))
end

@testset "kron / pattern_inverse / reshape preserve nnz" begin
    a = C.dense(_S, 2, 3); b = C.dense(_S, 4, 3)
    k = C.kron(a, b)
    @test (C.size1(k), C.size2(k)) == (C.size1(a) * C.size1(b), C.size2(a) * C.size2(b))
    @test C.nnz(k) == C.nnz(a) * C.nnz(b)
    inv = C.pattern_inverse(C.lower(_S, 4))      # 16 cells - 10 lower = 6
    @test C.nnz(inv) == 6
    r = C.reshape(C.dense(_S, 4, 5), 2, 10)
    @test (C.size1(r), C.size2(r)) == (2, 10)
    @test C.nnz(r) == 20
end

@testset "sparsity of an SX / MX expression" begin
    x = sym(SX, "x", 3)
    # elementwise square -> dense 3x1; jacobian wrt x is diagonal.
    e = C.times(x, x)
    @test C.nnz(C.sparsity(e)) == 3
    J = jacobian(e, x)
    sJ = C.sparsity(J)
    @test (C.size1(sJ), C.size2(sJ)) == (3, 3)
    @test C.nnz(sJ) == 3                          # diagonal Jacobian
    m = sym(MX, "m", 4)
    @test C.nnz(C.sparsity(m + m)) == 4
end

@testset "DM built on a custom Sparsity has matching shape" begin
    sp = C.triplet(_S, 4, 3, [0, 2, 3], [1, 2, 1])
    d = DM(sp, DM([5.0, 6.0, 7.0]))
    @test (C.size1(d), C.size2(d)) == (4, 3)
    @test C.nnz(d) == 3
    @test C.get_col(C.sparsity(d)) == C.get_col(sp)
end
